core/include/cubos/core/al/audio_context.hpp

```diff
@@ -0,0 +1,200 @@
+#include <string>
+#include <vector>
+
+#include <glm/glm.hpp>
+
+#include <cubos/core/api.hpp>
+
+namespace cubos::core::al
+{
+    namespace impl
+    {
+        class Buffer;
+        class Source;
+        class Listener;
+        class AudioDevice;
+    } // namespace impl
+
+    /// @brief Handle to an audio buffer.
+    /// @see impl::Buffer - audio buffer interface.
+    /// @see AudioContext::createBuffer()
+    /// @ingroup core-al
+    using Buffer = std::shared_ptr<impl::Buffer>;
+
+    /// @brief Handle to an audio source.
+    /// @see impl::Source - audio source interface.
+    /// @see AudioDevice::createSource()
+    /// @ingroup core-al
+    using Source = std::shared_ptr<impl::Source>;
+
+    /// @brief Handle to an audio listener.
+    /// @see impl::Listener - audio listener interface.
+    /// @see AudioDevice::createListener()
+    /// @ingroup core-al
+    using Listener = std::shared_ptr<impl::Listener>;
+
+    /// @brief Handle to an audio device.
+    /// @see impl::AudioDevice - audio device interface.
+    /// @see AudioContext::createDevice()
+    /// @ingroup core-al
+    using AudioDevice = std::shared_ptr<impl::AudioDevice>;
+
+    /// @brief Audio context that contains audio devices;
+    class CUBOS_CORE_API AudioContext
+    {
+    public:
+        AudioContext() = default;
+        virtual ~AudioContext() = default;
+
+        /// @brief Creates an audio context.
+        /// @return AudioContext, or nullptr on failure.
+        static std::shared_ptr<AudioContext> create();
+
+        /// @brief Enumerates the available devices.
+        /// @param[out] devices Vector to fill with the available devices.
+        virtual void enumerateDevices(std::vector<std::string>& devices) = 0;
+
+        /// @brief Creates a new audio device
+        /// @param specifier The specifier of the audio device, used to identify it
+        /// @param listenerCount The number of audio listener to be created by the device's engine.
+        /// @return Handle of the new device
+        virtual AudioDevice createDevice(unsigned int listenerCount,
+                                         const std::string& specifier = "MiniaudioDevice") = 0;
+
+        /// @brief Creates a new audio buffer.
+        /// @param data Data to be written to the buffer, either .wav, .mp3 or .flac.
+        /// @param datSize Size of the data to be written.
+        /// @return Handle of the new buffer.
+        virtual Buffer createBuffer(const void* data, size_t dataSize) = 0;
+    };
+
+    /// @brief Namespace to store the abstract types implemented by the audio device implementations.
+    namespace impl
+    {
+        /// @brief Abstract audio buffer.
+        class CUBOS_CORE_API Buffer
+        {
+        public:
+            virtual ~Buffer() = default;
+
+            /// @brief  Gets the length in seconds of the audio buffer.
+            /// @return Length in seconds of the audio buffer.
+            virtual size_t length() = 0;
```

Comment:
Should this be a float then, perhaps? Or do we only have a per-second resolution?